[5.2] Modified DatabaseTransactions to work with multiple connections

tpavlek opened the pull request:
[Resubmitting, accidentally submitted previous one against 5.1]

This feature is fully backwards-compatible.

If a user is using DatabaseTransactions but the tests involve multiple database connections, they can now override the connectionsToTransact() method and return an array of their database names.

I also added a protected property transactingConnections because if one needs to select some data that has been inserted into a table that is undergoing a transaction using \DB::connection($connectionName) will return a new connection instead, and the data won't be there. This seemed like a better solution to me than registering singletons, etc.

If this PR is accepted, I'd be happy to submit a subsequent pull to update the docs.

tpavlek committed:
Modified DatabaseTransactions to work with multiple connections

GrahamCampbell commented on src/Illuminate/Foundation/Testing/DatabaseTransactions.php:
```diff
@@ -4,12 +4,42 @@
 trait DatabaseTransactions
 {
+    /**
+     * This property will store references to the connections currently affected by the transaction. They are stored
```
cs

GrahamCampbell commented on src/Illuminate/Foundation/Testing/DatabaseTransactions.php:
```diff
@@ -4,12 +4,42 @@
+    protected $transactingConnections = [];
+
+    /**
+     * Get the names of the database connections that should be included in the transaction. The default value of empty
```
cs

tpavlek replied:
What exactly is the cs violation here? Style checks all pass, and there's nothing in the contribution guide except PSR-2.

GrahamCampbell replied:
Please look at every other phpdoc in the entire framework.

GrahamCampbell commented on src/Illuminate/Foundation/Testing/DatabaseTransactions.php:
```diff
@@ -4,12 +4,42 @@
+        $db = $this->app->make('db');
+
+        foreach ($this->connectionsToTransact() as $connectionName) {
+            $connection = $db->connection($connectionName);
```
don't do this extra assigment please

GrahamCampbell replied:
ping

tpavlek replied:
Do you want it to be 

```
$this->transactingConnections[$connectionName] = $db->connection($connectionName);
$this->transactingConnections[$connectionName]->beginTransaction();
```

If that's what you want I can do that, but that seems worse to me.

And it can't be done in a one liner since beginTransaction returns void and I need that assignment to the currently transacting array.

GrahamCampbell replied:
> $connectionName

can be shorter

> transactingConnections

Also can be renamed to be shorter.

tpavlek commented:
@GrahamCampbell I've addressed your feedback. If the docblocks still aren't good you'll need to be more specific because I seriously don't know what the complaint is.

GrahamCampbell commented on src/Illuminate/Foundation/Testing/DatabaseTransactions.php:
```diff
@@ -4,12 +4,42 @@
+            foreach ($this->connectionsToTransact() as $connectionName) {
+                $this->app->make('db')->connection($connectionName)->rollBack();
+            }
+
```
cs

tpavlek replied:
Friend, I appreciate the work you do for open source but you're *really* not making me want to help out **at all**.

What code style? The contribution guide only says to run php-cs-fixer, which I have. I follow PSR-2. Are you complaining about the space above and below the foreach? Since I'm supposed to "look at every other file in the framework" I looked at `Illuminate\Support\Collection` which spaces every foreach like this. Is that not what you want now?

You can either have a contribution guide that details what you actually want, or you can tell people what you want in the pulls, but when you do *neither* you just come off as a jerk.

GrahamCampbell replied:
I'm referring to the extra newline after the closing brace.

GrahamCampbell replied:
> Since I'm supposed to "look at every other file in the framework

That is correct. That's the most complete documentation of our code style. It we were to write it out formally, it would be massive and impossible to read.

GrahamCampbell commented on src/Illuminate/Foundation/Testing/DatabaseTransactions.php:
```diff
@@ -4,12 +4,36 @@
+    /**
+     * Retrieve the names of the database connections that should be included in the transaction.
+     *
+     * The default value of empty string will cause only the default database connection to be included in the transaction.
```
Please limit line length to 80.

GrahamCampbell commented on src/Illuminate/Foundation/Testing/DatabaseTransactions.php:
```diff
@@ -4,12 +4,36 @@
 trait DatabaseTransactions
 {
+    /**
+     * The connections currently affected by the transaction. Indexed by connection name.
```
Please limit this to one sentence per line.

GrahamCampbell replied:
> Indexed by connection name.

Infact, just drop that part please.

taylorotwell commented:
I can handle the DocBlocks. Don't worry about it.

GrahamCampbell commented on src/Illuminate/Foundation/Testing/DatabaseTransactions.php:
```diff
@@ -4,12 +4,36 @@
     {
-        $this->app->make('db')->beginTransaction();
+        foreach ($this->connectionsToTransact() as $name) {
+            $this->transacting[$name] = $this->app->make('db')->connection($name);
```
Might be an idea to resolve `$this->app->make('db')` once outside the loop for performance reasons.

GrahamCampbell commented on src/Illuminate/Foundation/Testing/DatabaseTransactions.php:
```diff
@@ -4,12 +4,36 @@
         $this->beforeApplicationDestroyed(function () {
-            $this->app->make('db')->rollBack();
+            foreach ($this->connectionsToTransact() as $name) {
+                $this->app->make('db')->connection($name)->rollBack();
```
Same.

tpavlek commented:
@taylorotwell thanks. I squashed the previous commits so there wasn't a bunch of useless "style fixes"

tpavlek committed:
Resovled db call outside of loop

tpavlek commented:
With @taylorotwell's help, are there any problems left with this?

GrahamCampbell commented:
Looks good to me (after Taylor cleans it up). Thanks for the contribution.

taylorotwell commented:
Why would calling DB::connection($connectionName) make a new instance? It should always serve the same instance if it has been created before.

tpavlek committed:
Removed transactingconnections array

tpavlek commented:
Yeah, you're right @taylorotwell. I just tested it again with the code that failed before - there was some other bug in my test causing the issue, which is now resolved. Updated the pull - it's all much simpler now :+1:
